File: automcp/adapters/pydantic.py
from typing import Any, Callable, Type
from pydantic import BaseModel
import asyncio
import contextlib
import io


def create_pydantic_adapter(
    agent_instance: Any,
    name: str,
    description: str,
    input_schema: Type[BaseModel],
) -> Callable:
    """
    Convert a Pydantic agent instance to an MCP tool, making it async.

    Args:
        agent_instance: The Pydantic agent instance
        name: The name of the MCP tool
        description: The description of the MCP tool
        input_schema: The Pydantic model class defining the input schema

    Returns:
        An awaitable callable function that can be used as an MCP tool
    """
    # Get the input schema fields to preserve the signature
    schema_fields = input_schema.model_fields
    
    # Create the parameter string for the function signature
    params_str = ", ".join(
        f"{field_name}: {field_info.annotation.__name__ if hasattr(field_info.annotation, '__name__') else 'Any'}"
        for field_name, field_info in schema_fields.items()
    )

    # Create the function body with proper async/await
    body_str = f"""async def run_agent({params_str}):
        # Create the input object
        input_data = input_schema({', '.join(f'{name}={name}' for name in schema_fields)})
        
        # Call the async run method
        with contextlib.redirect_stdout(io.StringIO()):
            result = await agent_instance.run(input_data.query)
        
        # Process the result
        if hasattr(result, 'data'):
            return result.data
        elif hasattr(result, 'raw'):
            return result.raw
        else:
            return result
    """

    # Create a namespace for the exec
    namespace = {
        "input_schema": input_schema,
        "agent_instance": agent_instance,
        "asyncio": asyncio,
        "contextlib": contextlib,
        "io": io,
    }
    
    # Execute the function definition
    exec(body_str, namespace)
    
    # Get the created function
    run_agent = namespace["run_agent"]
    
    # Add metadata
    run_agent.__name__ = name
    run_agent.__doc__ = description
    
    return run_agent

File: src/adapters/pydantic.ts
import { ToolResult } from './base.js';
import { ensureSerializable, extractParamsFromArgs, parseInputWithSchema, type SchemaLike } from './utils.js';

interface BaseModel {
  [key: string]: any;
  query?: string;
}

interface FieldInfo {
  annotation?: any;
  required?: boolean;
  default?: any;
}

interface AgentResult {
  data?: any;
  raw?: any;
}

interface PydanticAgent {
  run(query: string): Promise<AgentResult | any>;
}

/**
 * Convert a Pydantic agent instance to an MCP tool, making it async.
 */
export function createPydanticAdapter(
  agentInstance: PydanticAgent,
  name: string,
  description: string,
  inputSchema: SchemaLike
): (...args: any[]) => Promise<ToolResult> {
  const runAgent = async (...args: any[]): Promise<ToolResult> => {
    const kwargs = extractParamsFromArgs(args, inputSchema);
    const parsedInput = parseInputWithSchema(inputSchema, kwargs);

    // Redirect stdout equivalent (capture console output)
    const originalLog = console.log;
    const originalWarn = console.warn;
    const originalError = console.error;
    const outputBuffer: string[] = [];
    
    // Override console methods to capture output
    console.log = (...args) => outputBuffer.push(args.join(' '));
    console.warn = (...args) => outputBuffer.push(args.join(' '));
    console.error = (...args) => outputBuffer.push(args.join(' '));

    try {
      // Call the async run method
      const queryInput = parsedInput && typeof parsedInput === 'object'
        ? (parsedInput as Record<string, any>).query ?? parsedInput
        : parsedInput;
      const result = await agentInstance.run(
        typeof queryInput === 'string' ? queryInput : JSON.stringify(queryInput)
      );

      let output: any = result;

      if (result && typeof result === 'object') {
        if ('data' in result) {
          output = result.data;
        } else if ('raw' in result) {
          output = result.raw;
        }
      }

      const serialized = ensureSerializable(output);

      if (serialized && typeof serialized === 'object') {
        return {
          content: [{ type: 'text', text: JSON.stringify(serialized, null, 2) }],
          structuredContent: serialized
        };
      }

      return {
        content: [{ type: 'text', text: String(serialized) }]
      };
    } finally {
      // Restore original console methods
      console.log = originalLog;
      console.warn = originalWarn;
      console.error = originalError;
    }
  };
  
  // Add metadata
  Object.defineProperty(runAgent, 'name', { value: name });
  Object.defineProperty(runAgent, 'description', { value: description });
  
  return runAgent;
}

/**
 * Create a typed wrapper for Pydantic adapter with specific input schema
 */
export function createTypedPydanticAdapter<T extends BaseModel>(
  agentInstance: PydanticAgent,
  name: string,
  description: string,
  inputSchema: SchemaLike
): (input: T) => Promise<ToolResult> {
  const adapter = createPydanticAdapter(agentInstance, name, description, inputSchema);

  return async (input: T): Promise<ToolResult> => {
    return await adapter(input);
  };
}

/**
 * Helper function to create a Pydantic-like model class
 */
export function createModel<T extends Record<string, any>>(
  fields: Record<keyof T, FieldInfo>
) {
  return class PydanticModel implements BaseModel {
    [key: string]: any;

    constructor(data: Partial<T>) {
      Object.assign(this, data);
      
      // Validate required fields
      for (const [fieldName, fieldInfo] of Object.entries(fields)) {
        if (fieldInfo.required && !(fieldName in data)) {
          throw new Error(`Field '${fieldName}' is required`);
        }
        if (!(fieldName in data) && fieldInfo.default !== undefined) {
          this[fieldName] = fieldInfo.default;
        }
      }
    }
    
    static model_fields = fields;
  };
} 